design/cassandra/design.md

```diff
@@ -0,0 +1,109 @@
+* Despite those advantages, maintaining a Cassandra cluster is a highly manual and tiring task, which requires extensive familiarity with Cassandra's internals.
+* An operator will automate all the manual interventions that a human administrator normally has to perform and provide a production-ready database out-of-the-box.
+
+## Features
+
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
+## Design
+
+### Overview
+
+* This operator will use the pattern: `CRD` + `Controller` + `Sidecar`
+
+![operator-design-overview](media/operator_overview.jpg)
+
+### Cassandra-Kubernetes Mapping
+
+* The design of the Cassandra CRD will follow the Cassandra terminology as much as possible, in order to be familiar to existing users of Cassandra and not confuse new users studying the Cassandra docs.
+
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
+|     Node      |     Pod       |
+
+An example CRD would look something like this: 
+
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "0.1"
+  dataCenters:
+  - name: "dc-1"
+    racks:
+    - name: "rack-1"
+      replicas: 3
+      nodeSelector:  
+        dc: 1
+        rack: 1
+      storage:
+        size: "500Gi"
+        storageClassName: "fast"
+      resources:
+        requests:
+          cpu: "2000m"
+          memory: "4Gi"
+        limits:
+          cpu: "2000m"
+          memory: "4Gi"
+  backup:
+    storageType: "S3"
+    schedule: "*/30 * * * *"
+  image:
+    repository: "cassandra"
+    tag: "3.11.2"
+    pullPolicy: "IfNotPresent"
+  sidecarImage: 
+    repository: "rook.io/sidecar-cassandra"
+    tag: "0.1"
+```
+
+* The operator will create a StatefulSet for each (Datacenter, Rack) tupple.
+* In practice, Datacenter is usually mapped to a Region and Rack to an Availability Zone.
+
+
+### Sidecar
+
+In Cassandra, many actions require access to the cassandra process or the underlying system. More specifically:
+
+* *Cluster Creation:* dynamically insert java classes and other preparations for Cassandra to run. 
+* *Cassandra Node Replacement:* requires editing config files.
+* *Backup, Monitoring:* requires talking to a JMX interface that the Cassandra process exposes.
+
+By running a sidecar alongside Cassandra, we can monitor the process closely and deal with complex failure scenarios. 
```

Comment:
Can we get into some more details about the sidecar? What binaries/tools need to be in the sidecar container image?  How does it know when to do work (does it watch k8s API, does it expose a REST endpoint of its own, etc.)? How does it talk to the Cassandra node instance it's running along side, is that through exec'ing to a Cassandra standard tool, or does it talk a wire protocol, etc.?

Reply:
After much thought, I think a good approach would be for each sidecar to be a mini-controller responsible for its Cassandra instance. Each sidecar will have its own (Spec, Status), where Spec will be updated by the central controller and Status will be updated by the sidecar itself. That means the sidecar will include client-go and will watch the Kubernetes API. Since it will only watch its own CRD, I expect it to be lightweight. The communication with Cassandra will be done through a socket (Unix preferably, else TCP) using a JMX-HTTP bridge ([Jolokia](https://jolokia.org/)).

Reply:
Are you using sidecars as a control channel so that the central Rook operator can communicate with each instance of Cassandra running in the cluster?

Reply:
Essentially, I am using the Sidecar CRDs as a mean of communication between the central controller and each cassandra instance. This is in line with the Kubernetes level driven design and resembles the way the kubelet works. The kubelet watches etcd for Pods which have nodeName set to its node's name, in order to know which Pods it needs to create.